### PR TITLE
feat: Personalized authentication banner

### PR DESCRIPTION
**Pull request by AmarTrebinjac**
## Changes
TODO BEFORE MERGE:
- [x] Add back experiment check

Adds personalized authentication banner with order priority: User who shared > Social Media referrer > Geolocation

For review purposes, I commented out the experiment check. Will re-add before merging.

## Events

Did you introduce any new tracking events?

## Experiment

Yes, `post_banner_personalized`

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

MI-638 #done
MI-639 #done
MI-640 #done

### Preview domain
https://mi-638.preview.app.daily.dev

### CODE REVIEW

File: packages/shared/src/components/auth/AuthenticationBanner.tsx [hunk 1]
```diff
@@ -17,13 +17,19 @@ import {
...
+export function AuthenticationBanner({
+  children,
+}: AuthenticationBannerProps): ReactElement {
   const { showLogin } = useAuthContext();
 
   return (
     <BottomBannerContainer
       className={classNames(
         'gap-24 border-t border-accent-cabbage-default py-10 shadow-3 laptopL:gap-32',
+        authGradientBg,
```
Comment (AmarTrebinjac) [inline packages/shared/src/components/auth/AuthenticationBanner.tsx:32]: re-added this for cases where people are on slow internet speeds and the BG image loads slowly

File: packages/shared/src/components/auth/AuthenticationBanner.tsx [hunk 2]
```diff
@@ -17,13 +17,19 @@ import {
 
 const Section = classed('div', 'flex flex-col');
 
-export function AuthenticationBanner(): ReactElement {
+type AuthenticationBannerProps = {
+  children?: ReactNode | ReactNode[];
+};
+export function AuthenticationBanner({
+  children,
+}: AuthenticationBannerProps): ReactElement {
```
Comment (sshanzel) [inline packages/shared/src/components/auth/AuthenticationBanner.tsx:25]: I think we use just this so we won't have to introduce one. Just import this through React.

```suggestion
}: PropsWithChildren): ReactElement {
```

File: packages/shared/src/components/auth/AuthenticationBanner.tsx [hunk 3]
```diff
@@ -32,12 +38,18 @@ export function AuthenticationBanner(): ReactElement {
...
+      <Section className="w-[32.5rem] gap-4">
+        {children}
+        {!children && (
+          <OnboardingHeadline
+            className={{
+              title: 'typo-mega3',
+              description: 'mb-8 typo-title3',
+            }}
+          />
+        )}
```
Comment (sshanzel) [inline packages/shared/src/components/auth/AuthenticationBanner.tsx:50]: Would this not work? Not a blocker though.

```suggestion
        {children || (
          <OnboardingHeadline
            className={{
              title: 'typo-mega3',
              description: 'mb-8 typo-title3',
            }}
          />
        )}
```

File: packages/shared/src/components/auth/AuthenticationBanner.tsx [hunk 4]
```diff
@@ -32,12 +38,18 @@ export function AuthenticationBanner(): ReactElement {
...
       </Section>
-      <Section className="w-[23.25rem] pt-2">
+      <Section className="w-[23.25rem]">
```
Comment (sshanzel) [inline packages/shared/src/components/auth/AuthenticationBanner.tsx:52]: Was the change necessary here?

File: packages/shared/src/components/auth/OnboardingHeadline.tsx [hunk 5]
```diff
@@ -1,30 +1,57 @@
...
+export function OnboardingHeadline(
+  {
+    pretitle,
+    title = defaultTitle,
+    description = defaultDescription,
+    className,
+  }: OnboardingHeadlineProps = {
+    title: defaultTitle,
+    description: defaultDescription,
+  },
```
Comment (sshanzel) [inline packages/shared/src/components/auth/OnboardingHeadline.tsx:32]: We are setting the fallback value up above, would that not work alone?

```suggestion
  }: OnboardingHeadlineProps,
```

File: packages/shared/src/components/auth/OnboardingHeadline.tsx [hunk 6]
```diff
@@ -1,30 +1,57 @@
...
+): ReactElement {
+  return (
+    <div>
```
Comment (sshanzel) [inline packages/shared/src/components/auth/OnboardingHeadline.tsx:35]: Just to ensure it will always work as intended.

```suggestion
    <div className="flex flex-col">
```

File: packages/shared/src/graphql/users.ts [hunk 7]
```diff
@@ -660,3 +660,13 @@ export const TOP_READER_BADGE_BY_ID = gql`
 
   ${TOP_READER_BADGE_FRAGMENT}
 `;
+
+export const getBasicUserInfo = async (
```
Comment (sshanzel) [inline packages/shared/src/graphql/users.ts:664]: Great to wee we introduced this 🚀 

Should we wrap it with a try-catch? Then return null on fail?

File: packages/shared/src/lib/socialMedia.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+export const getSocialReferrer = (): string | null => {
+  if (!document.referrer) {
+    return null;
+  }
+
+  const url = new URL(document.referrer);
+  const host = url.hostname;
+
+  if (host.includes('reddit.')) {
+    return 'reddit';
```
Comment (sshanzel) [inline packages/shared/src/lib/socialMedia.ts:46]:
```suggestion
    return SocialIconType.Reddit;
```

File: packages/shared/src/lib/socialMedia.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  }
+  if (host.includes('x.') || host.includes('twitter.')) {
+    return 'x';
```
Comment (sshanzel) [inline packages/shared/src/lib/socialMedia.ts:49]:
```suggestion
    return SocialIconType.X;
```

File: packages/shared/src/lib/socialMedia.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    title: "You've mastered the art of finding hidden gems in the chaos.",
+    description:
+      "Now imagine a place where tech content is served fresh, minus the endless scrolls of cat memes (unless you're into that).",
+  },
+};
+
+/**
+ * NOTE! document.referrer does not contain a referrer on localhost
+ */
+export const getSocialReferrer = (): string | null => {
```
Comment (sshanzel) [inline packages/shared/src/lib/socialMedia.ts:37]:
```suggestion
export const getSocialReferrer = (): SocialIconType | null => {
```

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 11]
```diff
@@ -1,11 +1,110 @@
...
+          description: 'mb-8 typo-title3',
+        }}
+        title={`daily.dev is the fastest growing developer platform in ${country}!`}
+        description="We know how hard it is to be a developer. It doesn't have to be. Personalized news feed, dev community and search, much better than what's out there. Maybe ;)"
+      />
+    </AuthenticationBanner>
+  );
+};
+
+const SocialPersonalizedBanner = ({ site }: { site: string }): ReactElement => {
```
Comment (sshanzel) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:51]:
```suggestion
const SocialPersonalizedBanner = ({ site }: { site: SocialIconType }): ReactElement => {
```

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 12]
```diff
@@ -1,11 +1,110 @@
...
+  const Icon = socialIcon[site];
+  const gradient = socialGradient[site];
+  return (
+    <AuthenticationBanner>
+      <Icon size={IconSize.XXLarge} secondary={site === 'reddit'} />
```
Comment (sshanzel) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:56]:
```suggestion
      <Icon size={IconSize.XXLarge} secondary={site === SocialIconType.Reddit} />
```

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 13]
```diff
@@ -1,11 +1,110 @@
...
+      <OnboardingHeadline
+        className={{
+          title: `typo-mega3 ${gradient}`,
```
Comment (sshanzel) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:59]:
```suggestion
          title: classNames('typo-mega3', gradient),
```

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 14]
```diff
@@ -1,11 +1,110 @@
...
+        className={{
+          title: `typo-mega3`,
+          description: 'mb-8 typo-title3',
+        }}
+        title={`daily.dev is the fastest growing developer platform in ${country}!`}
+        description="We know how hard it is to be a developer. It doesn't have to be. Personalized news feed, dev community and search, much better than what's out there. Maybe ;)"
+      />
+    </AuthenticationBanner>
+  );
+};
```
Comment (sshanzel) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:49]: I think we can move along these components to their own separate file. Maybe just create a folder for example `banner` or `postBanner` to contain everything in there.

**Review by sshanzel — COMMENTED** [review]
Looking great so far! Just had a few minor comments but other than that, the implementation looked solid 💪

File: packages/shared/src/components/auth/AuthenticationBanner.tsx (hunk 2 above)
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/AuthenticationBanner.tsx:25]: Nice, didn't know this was a type!

File: packages/shared/src/components/auth/OnboardingHeadline.tsx (hunk 5 above)
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/OnboardingHeadline.tsx:32]: It does, oversight when refactoring :)

File: packages/shared/src/components/auth/AuthenticationBanner.tsx (hunk 4 above)
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/AuthenticationBanner.tsx:52]: Removing it lines up the icon/avatar relative to the input according to the design. Moves the text up a little when on the fallback one, but still looks correct to me. Could conditionally add the padding if no children.

File: packages/shared/src/graphql/users.ts (hunk 7 above)
Comment (AmarTrebinjac) [reply packages/shared/src/graphql/users.ts:664]: React query should handle that for you and return true on isError if it throws an error

File: packages/shared/src/components/auth/PostAuthBanner.tsx (hunk 14 above)
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:49]: @sshanzel I thought about it, but also didn't mind leaving them there because the PostAuthBanner doesn't even total 150 lines with all of them there. Don't see any other re-use for them currently.

File: packages/shared/src/graphql/users.ts (hunk 7 above)
Comment (sshanzel) [reply packages/shared/src/graphql/users.ts:664]: Oh, nice. Yeah, if we wrap it in useQuery, then we should not need to wrap this one.

File: packages/shared/src/components/auth/PostAuthBanner.tsx (hunk 14 above)
Comment (sshanzel) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:49]: Yeah, this was more of a non-blocking. Was just suggestion.

File: packages/shared/src/components/auth/PostAuthBanner.tsx (hunk 13 above)
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/PostAuthBanner.tsx]: I tried applying this just for the sake of testing if it would fix the gradient issue I mentioned on slack (it didnt 😭)

But is there any point in adding classNames here, when the component that receives the css classes already applies it, @sshanzel ?

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 15]
```diff
@@ -1,11 +1,116 @@
...
+  getSocialReferrer,
+  socialCTA,
+  socialGradient,
+  socialIcon,
+  type SocialIconType,
+} from '../../lib/socialMedia';
+import { capitalize } from '../../lib/strings';
+import { useAuthContext } from '../../contexts/AuthContext';
+
+const geoToEmoji = (geo: string): string => {
```
Comment (rebelchris) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:25]: It's fine for now, but we could have moved it to lib function assuming product might want this in more places

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 16]
```diff
@@ -1,11 +1,116 @@
...
+
+  return (
+    <AuthenticationBanner>
+      <span className="text-[3.5rem] leading-none">{emoji}</span>
+      <OnboardingHeadline
+        className={{
+          title: `typo-mega3`,
+          description: 'mb-8 typo-title3',
+        }}
+        title={`daily.dev is the fastest growing developer platform in ${country}!`}
```
Comment (rebelchris) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:46]: Is there a chance geo is empty?

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 17]
```diff
@@ -1,11 +1,116 @@
...
+};
+
+const GeoPersonalizedBanner = ({ geo }: { geo: string }): ReactElement => {
+  const emoji = geoToEmoji(geo);
+  const displayNames = new Intl.DisplayNames(['en'], { type: 'region' });
+  const country = displayNames.of(geo.toUpperCase());
+
+  return (
+    <AuthenticationBanner>
+      <span className="text-[3.5rem] leading-none">{emoji}</span>
```
Comment (rebelchris) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:40]: Rather use Typoghraph element here, it should be `typo-mega1` from design.

Screenshot 2024-11-25 at 08 46 23

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 18]
```diff
@@ -1,11 +1,116 @@
...
+  if (isError) {
+    return <AuthenticationBanner />;
+  }
+
+  const name = user?.name ? user?.name.split(' ')[0] : user?.username;
+
+  return (
+    <AuthenticationBanner>
+      {user?.image && (
+        <Image
```
Comment (rebelchris) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:94]: Should we use ProfileImage here? (or whatever it's called, it handles all classes already)

File: packages/shared/src/lib/socialMedia.ts [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+  X = 'x',
+}
+
+export const socialIcon: Record<SocialIconType, ComponentType<IconProps>> = {
+  [SocialIconType.Reddit]: RedditIcon,
+  [SocialIconType.X]: TwitterIcon,
+};
+
+export const socialGradient = {
+  reddit: 'bg-gradient-to-r !from-[#FF4500] !to-[#FF6B00]',
```
Comment (rebelchris) [inline packages/shared/src/lib/socialMedia.ts:17]: This is supposed to be custom I guess?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Nice, some questions mainly, overall looks great

File: packages/shared/src/components/auth/PostAuthBanner.tsx (hunk 16 above)
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:46]: @rebelchris There is, but we check in PostAuthBanner before going down this route:

`
 if (geo?.region) {
    return <GeoPersonalizedBanner geo={geo.region} />;
  }
  `

File: packages/shared/src/components/auth/PostAuthBanner.tsx (hunk 15 above)
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:25]: I'll move it, no problem!

File: packages/shared/src/lib/socialMedia.ts (hunk 19 above)
Comment (AmarTrebinjac) [reply packages/shared/src/lib/socialMedia.ts:17]: @rebelchris Yes, this is the color for the reddit ones. The importants are because of this weird issue: https://dailydotdev.slack.com/archives/C02E2C3C13R/p1732285583763759 🥴

File: packages/shared/src/components/auth/PostAuthBanner.tsx (hunk 16 above)
Comment (rebelchris) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:46]: Yeah sorry, saw it actually end of review all good here ✔️

File: packages/shared/src/components/auth/PostAuthBanner.tsx (hunk 15 above)
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:25]: @rebelchris Moved, and also moved "geoToCountry" and added an extra check on it. Realized there's quite a few countries with "The" prefix, such as The United States, The Netherlands, The United Kingdom etc. 

old:
Screenshot 2024-11-25 at 10 57 55

new: 
Screenshot 2024-11-25 at 10 45 47

File: packages/shared/src/components/auth/OnboardingHeadline.tsx [hunk 20]
```diff
@@ -1,30 +1,52 @@
...
-
-      <h2 className={className?.description}>
-        We know how hard it is to be a developer. It doesn&apos;t have to be.
-        Personalized news feed, dev community and search, much better than
-        what&apos;s out there. Maybe ;)
-      </h2>
-    </>
+    <div className="flex flex-col">
+      {pretitle && (
+        <p
```
Comment (sshanzel) [inline packages/shared/src/components/auth/OnboardingHeadline.tsx:32]: I think we can use Typography here.

File: packages/shared/src/components/auth/OnboardingHeadline.tsx [hunk 21]
```diff
@@ -1,30 +1,52 @@
...
+          className={classNames(
+            'font-bold text-white typo-mega3',
+            className?.pretitle,
+          )}
+        >
+          {pretitle}
+        </p>
+      )}
+      <h1
```
Comment (sshanzel) [inline packages/shared/src/components/auth/OnboardingHeadline.tsx:41]: I think we can use Typography here too.

File: packages/shared/src/components/auth/OnboardingHeadline.tsx [hunk 22]
```diff
@@ -1,30 +1,52 @@
...
+        className={classNames(
+          `mb-4 ${OnboardingGradientClasses}`,
+          className?.title,
```
Comment (sshanzel) [inline packages/shared/src/components/auth/OnboardingHeadline.tsx:44]:
```suggestion
          `mb-4`,
          className?.title,
          OnboardingGradientClasses,
```

File: packages/shared/src/components/auth/PostAuthBanner.tsx (hunk 15 above)
Comment (sshanzel) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:25]: @AmarTrebinjac I think changes have not been pushed yet.

Comment (sshanzel) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:25]: Ah, it has been pushed, just this one left.

**Review by sshanzel — COMMENTED** [review]
Last few things but nothing big! 🚢

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 23]
```diff
@@ -15,9 +120,30 @@ export const PostAuthBanner = (): ReactElement => {
...
+  // });
+
   if (showExtensionCTA) {
     return <AuthExtensionBanner />;
   }
 
+  // if (showPersonalizedBanner) {
+  const userId = searchParams.get('userid');
+
+  if (userId) {
```
[the remark this answers is not in the document]
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:53]: Since both you and @sshanzel mentioned it, I moved them all to separate components and dynamically load them

**Review by rebelchris — APPROVED** [review]
Nice looks good to me, please do bring back the experiment code before merging

File: packages/shared/src/components/auth/PostAuthBanner.tsx [hunk 24]
```diff
@@ -15,9 +38,30 @@ export const PostAuthBanner = (): ReactElement => {
     shouldEvaluate: isCompatibleBrowser,
   });
 
+  // const { value: showPersonalizedBanner } = useConditionalFeature({
+  //   feature: feature.postPersonalizedBanner,
+  //   shouldEvaluate: isCompatibleBrowser,
+  // });
```
Comment (sshanzel) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:44]: Since we are almost at the end stages of this, we can now revert the changes on this part.

File: packages/shared/src/components/banners/personalized/SocialPersonalizedBanner.tsx [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+const SocialPersonalizedBanner = ({
+  site,
+}: {
+  site: SocialIconType;
+}): ReactElement => {
+  const Icon = socialIcon[site];
+  const gradient = socialGradient[site];
+  return (
+    <AuthenticationBanner>
+      <Icon size={IconSize.XXLarge} secondary={site === 'reddit'} />
```
Comment (sshanzel) [inline packages/shared/src/components/banners/personalized/SocialPersonalizedBanner.tsx:22]:
```suggestion
      <Icon size={IconSize.XXLarge} secondary={site === SocialIconType.Reddit} />
```

File: packages/shared/src/components/banners/personalized/UserPersonalizedBanner.tsx [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+import { AuthenticationBanner, OnboardingHeadline } from '../../auth';
+import { ProfilePicture } from '../../ProfilePicture';
+
+const UserPersonalizedBanner = ({
+  userId,
+}: {
+  userId: string;
+}): ReactElement => {
+  const { data: user, isError } = useQuery({
+    queryKey: ['user_personalized_banner', userId],
```
Comment (sshanzel) [inline packages/shared/src/components/banners/personalized/UserPersonalizedBanner.tsx:13]: We have a function called `generateQueryKey`, kindly use that here to keep it uniform. If the key is not found yet, feel free to add it on the enum parameter.

**Review by sshanzel — CHANGES_REQUESTED** [review]
Most comments are non-blocking. Just requesting changes until hiding this behind an experiment is updated. Other than that, everything looked great 🚀

**Review by sshanzel — APPROVED** [review]
